Issue template and FAQ - autoscaler upstream sync (#188)
gardener/autoscaler: cluster-autoscaler/FAQ.md

# In the context of Gardener:

### How do I sync gardener autoscaler with an upstream autoscaler minor release?

This is helpful in order to offer Gardener CA with latest or recent K8s version. Note that this may also demand a need to upgrade K8s version used by Machine Controller Manager.

> Assumption: We assume that the developer executing the below stages wants to synchronize with a certain minor release `1.x.y` of the cluster autoscaler

Reviewer: The procedure to sync with an upstream patch release is a little different , lets add it later.
```suggestion
Assumption: We assume that the developer executing the below stages wants to synchronise with a certain minor release `1.x.0` of the cluster autoscaler
```

#### Stage A: Identify CA Release branch and Record Release commit id

> 1. Go to: [K8S Autoscaler Releases](https://github.com/kubernetes/autoscaler/releases) and navigate to the release page of the specific `1.x.y` release. Example: [cluster-autoscaler-1.26.0 Release Page](https://github.com/kubernetes/autoscaler/releases/tag/cluster-autoscaler-1.26.0).

Reviewer: ```suggestion
1. Go to: [K8S Autoscaler Releases](https://github.com/kubernetes/autoscaler/releases) and navigate to the release page of the specific `1.x.0` release. Example: [cluster-autoscaler-1.26.0 Release Page](https://github.com/kubernetes/autoscaler/releases/tag/cluster-autoscaler-1.26.0).
```

Author: fixed

1. Fork `github.com/gardener/autoscaler` into your github account named `user` such that you now have a forked repo `https://github.com/user/autoscaler`
1. Check out `github.com/gardener/autoscaler` under `$GOPATH/src/k8s.io`
1. Change to checked-out `autoscaler` dir and set `upstream`  and `user` (fork) targets
    1. `git remote add user https://github.com/user/autoscaler`
    1. `git remote add upstream https://github.com/kubernetes/autoscaler.git`
1. Execute: `git fetch –all`
1. Checkout the `releaseBranch`: `git checkout -b upstream-release-1.x.0 upstream/cluster-autoscaler-release-1.x.0`
1. Hard Reset to the `releaseCommitId`: `git reset --hard releaseCommitId`
1. Check out and pull the primary branch for gardener fork which is not `master`/`main` but is instead named `machine-controller-manager-provider`:  
   1. `git checkout machine-controller-manager-provider`
   1. `git pull origin machine-controller-manager-provider`

#### Stage C:  Create Sync Branch, Perform Merge, Fix Vendor
1. Create a sync branch: `git checkout -b sync-upstream-v1.x.0`. For example, if you are syncing against `1.26.0`, this would be: `git checkout -b sync-upstream-v1.26.0`

> 1. Merge against the `releaseBranch` (which you had hard-reset to the `releaseCommitId` earlier): `git merge upstream-release-1.2x`

Reviewer: ```suggestion
1. Merge against the `releaseBranch` (which you had hard-reset to the `releaseCommitId` earlier): `git merge upstream-release-1.x.0`
```

Author: fixed

   - This is where changes of master get merged in. advantage of merging is it won’t change the commit hashes of already existing commits.
   - Accept all vpa changes.
   - Accept our changes in `go.mod` and `go.sum` and `vendor` directory.  (The `hack/update-vendor.sh` script which will be executed in later step expected to fix things)
1. In `cluster-autoscaler/go.mod`, upgrade versions of `machine-controller-manager-provider-aws`,  `machine-controller-manager-provider-azure` to the latest available release.

> 1. Run update vendor script after changing to `cluster-autoscaler` directory:  `./hack/update-vendor.sh 1.x.y` 

Reviewer: ```suggestion
Run update vendor script after changing to `cluster-autoscaler` directory: `./hack/update-vendor.sh 1.x.0`
```

Author: fixed

   - If the above still gives test issues then use `rsync` or `diff -rq` to figure out differences in `vendor` directory between upstream and our fork and synchronize them manually.

> 1. Create a new file `cluster-autoscaler/SYNC-CHANGES/SYNC_CHANGES-1.x.y.md` summarily describing the changes done. Follow existing convention for sync changes. Use upstream release notes as a guide when needed.

Reviewer: we don't create this file for every `1.x.y` but only for `1.x`.
```suggestion
1. Create a new file `cluster-autoscaler/SYNC-CHANGES/SYNC_CHANGES-1.x.md` summarily describing the changes done. Follow existing convention for sync changes. Use upstream release notes as a guide when needed.
```

Author: fixed

#### Stage D: Verification
1. Run Core Autoscaler Unit tests: `cd cluster-autoscaler; go test $(go list ./... | grep -v cloudprovider | grep -v vendor | grep -v integration)` 
1. Run MCM cloud provider implementation tests: `go test $(go list ./cloudprovider/mcm/... | grep -v vendor)`
1. Verify that binary can be created using: `../.ci/build`

> 1. Verify that binary can be created using: `../.ci/build`

Reviewer: this line is repeating
```suggestion
```

Author: fixed

#### Stage C:  Create Sync Branch, Perform Merge, Fix Vendor
1. Create a sync branch: `git checkout -b sync-upstream-v1.x.0`. For example, if you are syncing against `1.26.0`, this would be: `git checkout -b sync-upstream-v1.26.0`
1. Merge against the `releaseBranch` (which you had hard-reset to the `releaseCommitId` earlier): `git merge upstream-release-1.x.0`
   - This is where changes of master get merged in. advantage of merging is it won’t change the commit hashes of already existing commits.
   - Accept all vpa changes.
   - Accept our changes in `go.mod` and `go.sum` and `vendor` directory.  (The `hack/update-vendor.sh` script which will be executed in later step expected to fix things)
1. In `cluster-autoscaler/go.mod`, upgrade versions of `machine-controller-manager-provider-aws`,  `machine-controller-manager-provider-azure` to the latest available release.
1. Run update vendor script after changing to `cluster-autoscaler` directory:  `./hack/update-vendor.sh 1.x.0` 
   - If the above still gives test issues then use `rsync` or `diff -rq` to figure out differences in `vendor` directory between upstream and our fork and synchronize them manually.

> 1. Create a new file `cluster-autoscaler/SYNC-CHANGES/SYNC_CHANGES-1.x.0.md` summarily describing the changes done. Follow existing convention for sync changes. Use upstream release notes as a guide when needed.

Reviewer: kindly commit this directly. If there are many such suggestions you could also add them to batch and then commit them from the UI itself.
```suggestion
1. Create a new file `cluster-autoscaler/SYNC-CHANGES/SYNC_CHANGES-1.x.md` summarily describing the changes done. Follow existing convention for sync changes. Use upstream release notes as a guide when needed.
```

Author: Hi Himanshu, this suggestion is outdated as the change was already made. ie file name is already  `cluster-autoscaler/SYNC-CHANGES/SYNC_CHANGES-1.x.md`
